src/java.base/share/classes/java/lang/reflect/Method.java

```diff
@@ -95,6 +95,8 @@ public final class Method extends Executable {
     // If this branching structure would ever contain cycles, deadlocks can
     // occur in annotation code.
     private Method              root;
+    // Hash code of this object
+    private int                 hash;
```

Comment:
We can declare this field `@Stable`, like the `methodAccessor`.

Reply:
This was something that I tried and I observed a performance regression on the ARM platform that I was testing. From my understanding, `@Stable` tells the compiler to trust the contents of this field which is only given when the field holder is a known constant. This is generally true for cases like `Enums` but not usually for `Method`. 

```
Benchmark                         Mode  Cnt  Score   Error  Units
# Intel Skylake
MethodHashCode.benchmarkHashCode  avgt    5  1.113 ± 1.146  ns/op
# Arm Neoverse N1
MethodHashCode.benchmarkHashCode  avgt    5  3.204 ± 0.001  ns/op
```

Reply:
Interesting, don't know about hotspot internals so I can't diagnose the exact cause of this regression.

Reply:
In order to be eligible for constant folding, the benchmark must declare the `Method hashCodeMethod;` as `static final`.

It is hard for me to understand why a `@Stable` annotation should have a detrimental performance impact on an instance field.

Can we see a benchmark on Arm Neoverse N1 with the field declared `@Stable` compared to not declared `@Stable`? Also, if the field is `static final`, how would it look like?

Reply:
As a note, If we would have access to the contemplated `StableValue` and `hash` was declared even as an _instance variable_ `StableValue<Integer>` in a regular class, then it would be trusted and would be eligible for constant folding due to the VM will have special rules for  fields of StableValue type.

Reply:
Ahh. There was a benchmark in the initial message. Sorry.

Reply:
We think this is caused by the membars being conservatively emitted at method exit, and more precis membar control should help ARM. Filed: https://bugs.openjdk.org/browse/JDK-8333791